File: client-pool.js
const { MetaAIClient } = require("./meta-client");

const HEALTH_CHECK_INTERVAL = 60_000; // 60 seconds
const STAGGER_DELAY = 5_000; // 5 seconds between warmup batches
const WARMUP_BATCH_SIZE = 2; // clients initialized in parallel per batch
const DEFAULT_QUEUE_TIMEOUT = 120_000; // 2 minutes
const SESSION_MAX_AGE = 10 * 60_000; // 10 minutes — refresh before tokens go stale

class ClientPool {
  /**
   * @param {number} poolSize Number of MetaAIClient instances to maintain
   * @param {object} [opts]
   * @param {number} [opts.queueTimeout] Max ms a request waits in queue (default 120s)
   */
  constructor(poolSize = 15, opts = {}) {
    this.poolSize = poolSize;
    this.queueTimeout = opts.queueTimeout || DEFAULT_QUEUE_TIMEOUT;
    this.proxy = opts.proxy || null;
    this.browserProxy = opts.browserProxy || null;
    this.entries = [];
    this.healthCheckTimer = null;
    this._waitQueue = []; // requests waiting for a free client
    this._completedCount = 0; // requests completed since last summary
    this._queuedCount = 0; // total queued (for log throttle)

    for (let i = 0; i < poolSize; i++) {
      this.entries.push({
        id: i,
        client: new MetaAIClient(this.proxy, i, { browserProxy: this.browserProxy }),
        busy: false,
        ready: false,
        initializing: false,
        sessionCreatedAt: null, // track when session was established
      });
    }
  }

  /**
   * Initialize clients in parallel batches with a stagger delay between batches.
   * Much faster than fully sequential for large pools.
   * @returns {Promise<void>}
   */
  async warmup() {
    console.log(`[pool] Warming up ${this.poolSize} clients (batches of ${WARMUP_BATCH_SIZE})...`);
    const startTime = Date.now();
    let readyCount = 0;

    for (let batchStart = 0; batchStart < this.entries.length; batchStart += WARMUP_BATCH_SIZE) {
      const batchEnd = Math.min(batchStart + WARMUP_BATCH_SIZE, this.entries.length);
      const batch = this.entries.slice(batchStart, batchEnd);
      const batchNum = Math.floor(batchStart / WARMUP_BATCH_SIZE) + 1;
      const totalBatches = Math.ceil(this.entries.length / WARMUP_BATCH_SIZE);

      console.log(`[pool] Batch ${batchNum}/${totalBatches}: initializing clients ${batchStart}-${batchEnd - 1}`);

      // Initialize all clients in this batch in parallel
      const results = await Promise.allSettled(
        batch.map(async (entry) => {
          entry.initializing = true;
          try {
            await entry.client.ensureSession();
            entry.ready = true;
            entry.sessionCreatedAt = Date.now();
            readyCount++;
            console.log(`[pool] Client ${entry.id} ready (${readyCount}/${this.poolSize})`);
          } catch (err) {
            console.error(`[pool] Client ${entry.id} warmup failed:`, err.message);
            entry.ready = false;
          } finally {
            entry.initializing = false;
          }
        })
      );

      // Stagger delay between batches (skip after last batch)
      if (batchEnd < this.entries.length) {
        await new Promise((resolve) => setTimeout(resolve, STAGGER_DELAY));
      }
    }

    if (readyCount === 0) {
      throw new Error("[pool] No clients could be initialized");
    }

    const elapsed = ((Date.now() - startTime) / 1000).toFixed(1);
    console.log(`[pool] Warmup complete: ${readyCount}/${this.poolSize} clients ready in ${elapsed}s`);

    // Start background health check
    this.healthCheckTimer = setInterval(() => this._healthCheck(), HEALTH_CHECK_INTERVAL);
    this.healthCheckTimer.unref();
  }

  /**
   * Acquire a free (ready + not busy) client.
   * If none available, wait in queue with a timeout.
   * @param {number} [timeoutMs] Override queue timeout for this request
   * @returns {Promise<object>} Pool entry
   */
  async acquire(timeoutMs) {
    const timeout = timeoutMs || this.queueTimeout;

    // Try to find a free client immediately
    const free = this.entries.find((e) => e.ready && !e.busy);
    if (free) {
      free.busy = true;
      return free;
    }

    // No free client — wait in queue
    const readyCount = this.entries.filter((e) => e.ready).length;
    if (readyCount === 0) {
      throw new Error("[pool] No healthy clients available");
    }

    this._queuedCount++;
    // Only log every 10th queued request to reduce spam
    if (this._queuedCount % 10 === 1 || this._queuedCount <= 3) {
      console.log(`[pool] All ${readyCount} clients busy, queue depth: ${this._waitQueue.length + 1}`);
    }

    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        // Remove from queue on timeout
        const idx = this._waitQueue.findIndex((w) => w.resolve === resolve);
        if (idx !== -1) this._waitQueue.splice(idx, 1);
        reject(new Error(`[pool] Queue timeout after ${timeout}ms — all clients busy`));
      }, timeout);

      this._waitQueue.push({ resolve, timer });
    });
  }

  /**
   * Release a client back to the pool. If requests are queued, hand it off.
   * @param {object} entry Pool entry returned by acquire()
   */
  release(entry) {
    if (this._waitQueue.length > 0 && entry.ready) {
      const next = this._waitQueue.shift();
      clearTimeout(next.timer);
      next.resolve(entry); // entry stays busy
    } else {
      entry.busy = false;
    }

    // Log a summary periodically instead of per-request
    this._completedCount++;
    if (this._completedCount % 10 === 0) {
      const busy = this.entries.filter((e) => e.busy).length;
      const queue = this._waitQueue.length;
      console.log(`[pool] Progress: ${this._completedCount} completed, ${busy} busy, ${queue} queued`);
    }
  }

  /**
   * Execute a function with an acquired client (1 request per client at a time).
   * @param {function} fn Async function receiving the MetaAIClient instance
   * @param {number} [timeoutMs] Optional queue timeout override
   * @returns {Promise<*>} Result of fn
   */
  async execute(fn, timeoutMs) {
    const entry = await this.acquire(timeoutMs);
    try {
      const result = await fn(entry.client);
      return result;
    } catch (err) {
      // If it looks like a session/auth error, mark client as unhealthy
      const msg = err.message || "";
      if (
        msg.includes("access token") ||
        msg.includes("LSD token") ||
        msg.includes("401") ||
        msg.includes("403") ||
        msg.includes("session") ||
        msg.includes("exhausted")
      ) {
        console.log(`[pool] Client ${entry.id} session error, marking unhealthy`);
        entry.ready = false;
        entry.sessionCreatedAt = null;
        entry.client.resetSession();
      }
      throw err;
    } finally {
      this.release(entry);
    }
  }

  /**
   * Background health check:
   * 1. Re-initialize any failed clients
   * 2. Proactively refresh idle clients with stale sessions (>10 min)
   */
  async _healthCheck() {
    // Re-init unhealthy clients
    const unhealthy = this.entries.filter(
      (e) => !e.ready && !e.initializing && !e.busy
    );

    if (unhealthy.length > 0) {
      console.log(`[pool] Health check: ${unhealthy.length} client(s) need re-init`);
      for (const entry of unhealthy) {
        try {
          entry.initializing = true;
          entry.client.resetSession();
          await entry.client.ensureSession();
          entry.ready = true;
          entry.sessionCreatedAt = Date.now();
          console.log(`[pool] Client ${entry.id} re-initialized`);
        } catch (err) {
          console.error(`[pool] Client ${entry.id} re-init failed:`, err.message);
        } finally {
          entry.initializing = false;
        }
      }
    }

    // Proactive token refresh: refresh idle clients with sessions older than 10 min
    const now = Date.now();
    const stale = this.entries.filter(
      (e) =>
        e.ready &&
        !e.busy &&
        !e.initializing &&
        e.sessionCreatedAt &&
        now - e.sessionCreatedAt > SESSION_MAX_AGE
    );

    if (stale.length > 0) {
      console.log(`[pool] Proactive refresh: ${stale.length} client(s) with stale sessions`);
      for (const entry of stale) {
        // Only refresh if still idle (another request may have grabbed it)
        if (entry.busy) continue;
        try {
          entry.initializing = true;
          entry.busy = true; // prevent acquire() from picking it up mid-refresh
          entry.client.resetSession();
          await entry.client.ensureSession();
          entry.sessionCreatedAt = Date.now();
          console.log(`[pool] Client ${entry.id} session refreshed`);
        } catch (err) {
          console.error(`[pool] Client ${entry.id} refresh failed:`, err.message);
          entry.ready = false;
          entry.sessionCreatedAt = null;
        } finally {
          entry.initializing = false;
          entry.busy = false;
        }
      }
    }
  }

  /**
   * Stop the health check timer and return pool stats.
   */
  shutdown() {
    if (this.healthCheckTimer) {
      clearInterval(this.healthCheckTimer);
      this.healthCheckTimer = null;
    }
    const ready = this.entries.filter((e) => e.ready).length;
    console.log(`[pool] Shutdown. ${ready}/${this.poolSize} clients were healthy. ${this._completedCount} requests served.`);
  }
}

module.exports = { ClientPool };

File: python/client_pool.py
"""
Client pool manager — maintains a pool of MetaAIClient instances with
warmup, acquire/release, health checks, and proactive session refresh.
"""

import asyncio
import time
from meta_client import MetaAIClient

HEALTH_CHECK_INTERVAL = 60  # seconds
STAGGER_DELAY = 5  # seconds between warmup batches
WARMUP_BATCH_SIZE = 2  # clients initialized in parallel per batch
DEFAULT_QUEUE_TIMEOUT = 120  # seconds
SESSION_MAX_AGE = 10 * 60  # 10 minutes


class ClientPool:
    def __init__(
        self,
        pool_size: int = 15,
        queue_timeout: float = DEFAULT_QUEUE_TIMEOUT,
        proxy: str | None = None,
    ):
        self.pool_size = pool_size
        self.queue_timeout = queue_timeout
        self.proxy = proxy
        self.entries: list[dict] = []
        self._health_check_task: asyncio.Task | None = None
        self._wait_queue: asyncio.Queue = asyncio.Queue()
        self._completed_count = 0
        self._queued_count = 0

        for i in range(pool_size):
            self.entries.append(
                {
                    "id": i,
                    "client": MetaAIClient(proxy=self.proxy, client_id=i),
                    "busy": False,
                    "ready": False,
                    "initializing": False,
                    "session_created_at": None,
                    "release_event": asyncio.Event(),
                }
            )

    async def warmup(self):
        """Initialize clients in parallel batches with a stagger delay between batches."""
        print(
            f"[pool] Warming up {self.pool_size} clients "
            f"(batches of {WARMUP_BATCH_SIZE})..."
        )
        start_time = time.time()
        ready_count = 0

        for batch_start in range(0, len(self.entries), WARMUP_BATCH_SIZE):
            batch_end = min(batch_start + WARMUP_BATCH_SIZE, len(self.entries))
            batch = self.entries[batch_start:batch_end]
            batch_num = batch_start // WARMUP_BATCH_SIZE + 1
            total_batches = (len(self.entries) + WARMUP_BATCH_SIZE - 1) // WARMUP_BATCH_SIZE

            print(
                f"[pool] Batch {batch_num}/{total_batches}: "
                f"initializing clients {batch_start}-{batch_end - 1}"
            )

            async def init_client(entry):
                nonlocal ready_count
                entry["initializing"] = True
                try:
                    await entry["client"].ensure_session()
                    entry["ready"] = True
                    entry["session_created_at"] = time.time()
                    ready_count += 1
                    print(
                        f"[pool] Client {entry['id']} ready "
                        f"({ready_count}/{self.pool_size})"
                    )
                except Exception as err:
                    print(f"[pool] Client {entry['id']} warmup failed: {err}")
                    entry["ready"] = False
                finally:
                    entry["initializing"] = False

            await asyncio.gather(
                *[init_client(entry) for entry in batch],
                return_exceptions=True,
            )

            # Stagger delay between batches (skip after last batch)
            if batch_end < len(self.entries):
                await asyncio.sleep(STAGGER_DELAY)

        if ready_count == 0:
            raise Exception("[pool] No clients could be initialized")

        elapsed = time.time() - start_time
        print(
            f"[pool] Warmup complete: {ready_count}/{self.pool_size} "
            f"clients ready in {elapsed:.1f}s"
        )

        # Start background health check
        self._health_check_task = asyncio.create_task(self._health_check_loop())

    async def acquire(self, timeout: float | None = None) -> dict:
        """
        Acquire a free (ready + not busy) client.
        If none available, wait with a timeout.
        """
        timeout = timeout or self.queue_timeout

        # Try to find a free client immediately
        for entry in self.entries:
            if entry["ready"] and not entry["busy"]:
                entry["busy"] = True
                return entry

        # No free client — wait for one to be released
        ready_count = sum(1 for e in self.entries if e["ready"])
        if ready_count == 0:
            raise Exception("[pool] No healthy clients available")

        self._queued_count += 1
        if self._queued_count % 10 == 1 or self._queued_count <= 3:
            print(
                f"[pool] All {ready_count} clients busy, "
                f"waiting for release..."
            )

        # Wait for any client to become free
        deadline = time.time() + timeout
        while True:
            remaining = deadline - time.time()
            if remaining <= 0:
                raise Exception(
                    f"[pool] Queue timeout after {timeout}s — all clients busy"
                )

            # Check again for a free client
            for entry in self.entries:
                if entry["ready"] and not entry["busy"]:
                    entry["busy"] = True
                    return entry

            # Wait a short interval and retry
            await asyncio.sleep(0.1)

    def release(self, entry: dict):
        """Release a client back to the pool."""
        entry["busy"] = False

        self._completed_count += 1
        if self._completed_count % 10 == 0:
            busy = sum(1 for e in self.entries if e["busy"])
            print(
                f"[pool] Progress: {self._completed_count} completed, "
                f"{busy} busy"
            )

    async def execute(self, fn, timeout: float | None = None):
        """
        Execute an async function with an acquired client.
        One request per client at a time.
        """
        entry = await self.acquire(timeout)
        try:
            result = await fn(entry["client"])
            return result
        except Exception as err:
            # If it looks like a session/auth error, mark client as unhealthy
            msg = str(err)
            if any(
                keyword in msg
                for keyword in [
                    "access token",
                    "LSD token",
                    "401",
                    "403",
                    "session",
                    "exhausted",
                ]
            ):
                print(
                    f"[pool] Client {entry['id']} session error, marking unhealthy"
                )
                entry["ready"] = False
                entry["session_created_at"] = None
                entry["client"].reset_session()
            raise
        finally:
            self.release(entry)

    async def _health_check_loop(self):
        """Run health checks periodically."""
        while True:
            await asyncio.sleep(HEALTH_CHECK_INTERVAL)
            try:
                await self._health_check()
            except Exception as err:
                print(f"[pool] Health check error: {err}")

    async def _health_check(self):
        """
        Background health check:
        1. Re-initialize any failed clients
        2. Proactively refresh idle clients with stale sessions (>10 min)
        """
        # Re-init unhealthy clients
        unhealthy = [
            e
            for e in self.entries
            if not e["ready"] and not e["initializing"] and not e["busy"]
        ]

        if unhealthy:
            print(f"[pool] Health check: {len(unhealthy)} client(s) need re-init")
            for entry in unhealthy:
                try:
                    entry["initializing"] = True
                    entry["client"].reset_session()
                    await entry["client"].ensure_session()
                    entry["ready"] = True
                    entry["session_created_at"] = time.time()
                    print(f"[pool] Client {entry['id']} re-initialized")
                except Exception as err:
                    print(f"[pool] Client {entry['id']} re-init failed: {err}")
                finally:
                    entry["initializing"] = False

        # Proactive token refresh for stale sessions
        now = time.time()
        stale = [
            e
            for e in self.entries
            if e["ready"]
            and not e["busy"]
            and not e["initializing"]
            and e["session_created_at"]
            and now - e["session_created_at"] > SESSION_MAX_AGE
        ]

        if stale:
            print(
                f"[pool] Proactive refresh: {len(stale)} client(s) with stale sessions"
            )
            for entry in stale:
                if entry["busy"]:
                    continue
                try:
                    entry["initializing"] = True
                    entry["busy"] = True
                    entry["client"].reset_session()
                    await entry["client"].ensure_session()
                    entry["session_created_at"] = time.time()
                    print(f"[pool] Client {entry['id']} session refreshed")
                except Exception as err:
                    print(f"[pool] Client {entry['id']} refresh failed: {err}")
                    entry["ready"] = False
                    entry["session_created_at"] = None
                finally:
                    entry["initializing"] = False
                    entry["busy"] = False

    async def shutdown(self):
        """Stop the health check and close all client sessions."""
        if self._health_check_task:
            self._health_check_task.cancel()
            try:
                await self._health_check_task
            except asyncio.CancelledError:
                pass
            self._health_check_task = None

        ready = sum(1 for e in self.entries if e["ready"])
        print(
            f"[pool] Shutdown. {ready}/{self.pool_size} clients were healthy. "
            f"{self._completed_count} requests served."
        )

        # Close all client sessions
        for entry in self.entries:
            try:
                await entry["client"].close()
            except Exception:
                pass
